File: sentinel/engine-py/sentinel/m4_trend.py
"""M4 — Trend & Trajectory Engine. Windowed feature computations consumed by
M5, plus content-rule trajectory classification. Refuses to emit a slope from
fewer than min_points (§3 M4).
"""
from __future__ import annotations

from .canonical import fmt_val, q6
from .stats import median, ols_slope


class Features:
    """Deterministic windowed views over usable observations.

    Usable = valid | suspect | mechanism-protected artifact (DECISIONS.md D7).
    Ordinal metrics (from content enums) are held as (index, label) pairs.
    """

    def __init__(self, accepted, content, reference_time, flags=None):
        self.ref = reference_time
        self.enums = content["tables"]["operational_bounds"]["enums"]
        self.flags = flags if flags is not None else set()
        screening = content["tables"]["stream_screening"]
        order = screening["source_priority"]
        self.by_metric: dict = {}
        self.events: list = []
        for o in accepted:  # already sorted by (ts, obs_id)
            usable = o["quality"] != "artifact_likely" or o.get("mechanism_protected")
            if o["type"] == "event":
                self.events.append(o)
                continue
            if not usable or "value" not in o:
                continue
            prio = order.index(o["source"]) if o["source"] in order else len(order)
            if o["type"] in self.enums:
                idx = self.enums[o["type"]].index(o["value"])
                point = (o["ts"], idx, o["value"], o["stream"], prio)
            else:
                point = (o["ts"], o["value"], None, o["stream"], prio)
            self.by_metric.setdefault(o["type"], []).append(point)

    def ordinal_index(self, metric, label):
        order = self.enums.get(metric)
        if order is None or label not in order:
            return None
        return order.index(label)

    def in_window(self, metric, window_min):
        lo = self.ref - window_min * 60
        return [p for p in self.by_metric.get(metric, []) if lo <= p[0] <= self.ref]

    def _select(self, metric, window_min, min_needed):
        """Stream fusion (DECISIONS.md D19): use the highest-trust single
        stream that can answer the question on its own. Pooling readings from
        different devices is a last resort — inter-device offsets masquerade
        as trends — and is always flagged, never silent."""
        pts = self.in_window(metric, window_min)
        if not pts:
            return pts
        groups: dict = {}
        for p in pts:
            groups.setdefault((p[4], p[3]), []).append(p)
        for key in sorted(groups):
            if len(groups[key]) >= min_needed:
                return groups[key]
        if len(groups) > 1 and len(pts) >= min_needed:
            self.flags.add("pooled_streams")
            return pts
        return pts

    def latest(self, metric, window_min):
        pts = self._select(metric, window_min, 1)
        return pts[-1] if pts else None

    def window_median(self, metric, window_min, min_points):
        pts = self._select(metric, window_min, min_points)
        if len(pts) < min_points:
            return None
        return median([p[1] for p in pts])

    def slope(self, metric, window_min, min_points):
        pts = self._select(metric, window_min, min_points)
        return ols_slope([(p[0], p[1]) for p in pts], min_points)

    def sustained_min(self, metric, window_min, threshold, direction):
        pts = self._select(metric, window_min, 1)
        if not pts:
            return None
        def beyond(v):
            return v > threshold if direction == "above" else v < threshold
        if not beyond(pts[-1][1]):
            return 0.0
        start = len(pts) - 1
        while start > 0 and beyond(pts[start - 1][1]):
            start -= 1
        return q6((pts[-1][0] - pts[start][0]) / 60.0)

    def crossings(self, metric, window_min, threshold, direction):
        pts = self._select(metric, window_min, 2)
        if len(pts) < 2:
            return None
        count = 0
        for i in range(1, len(pts)):
            prev_v, cur_v = pts[i - 1][1], pts[i][1]
            if direction == "above" and prev_v <= threshold < cur_v:
                count += 1
            elif direction == "below" and prev_v >= threshold > cur_v:
                count += 1
        return count

    def event_present(self, event_id, window_min):
        lo = self.ref - window_min * 60
        return any(e["event_id"] == event_id and lo <= e["ts"] <= self.ref for e in self.events)


def classify_trajectory(features, content, trace):
    rules = content["tables"]["trajectory_rules"]
    worse = improving = evaluated = 0
    details = []
    for m in rules["metrics"]:
        s = features.slope(m["metric"], rules["window_min"], rules["min_points"])
        if s is None:
            continue
        evaluated += 1
        thr = m["slope_per_min"]
        if m["worse_direction"] == "up":
            if s >= thr:
                worse += 1
                details.append(f"{m['metric']} worsening (slope {fmt_val(s)}/min)")
            elif s <= -thr:
                improving += 1
        else:
            if s <= -thr:
                worse += 1
                details.append(f"{m['metric']} worsening (slope {fmt_val(s)}/min)")
            elif s >= thr:
                improving += 1
    if evaluated == 0:
        result = "unknown"
    elif worse >= rules["worsening_min_signals"]:
        result = "worsening"
    elif improving >= rules["improving_min_signals"] and worse == 0:
        result = "improving"
    else:
        result = "stable"
    detail = f"trajectory {result} (worse_signals={worse}, improving_signals={improving}, metrics_evaluated={evaluated})"
    if details:
        detail += ": " + "; ".join(details)
    trace.append({"stage": "M4", "detail": detail})
    return result

File: sentinel/engine-ts/src/m4_trend.ts
/** M4 — Trend & Trajectory Engine. Windowed feature computations consumed by
 * M5, plus content-rule trajectory classification. Refuses to emit a slope
 * from fewer than min_points (§3 M4).
 */
import { fmtVal, own, q6 } from "./canonical.js";
import type { ContentHandle } from "./content.js";
import type { NormalizedObs } from "./m1_ingest.js";
import { median, olsSlope } from "./stats.js";

/** (ts, raw-or-ordinal-index, ordinal-label-or-null, stream, priority) — mirrors the Python tuple. */
export type Point = [number, number, string | null, string, number];

export class Features {
  ref: number;
  enums: Record<string, string[]>;
  byMetric: Map<string, Point[]> = new Map();
  events: NormalizedObs[] = [];

  flags: Set<string>;

  constructor(accepted: NormalizedObs[], content: ContentHandle, referenceTime: number, flags?: Set<string>) {
    this.ref = referenceTime;
    this.enums = content.tables["operational_bounds"].enums;
    this.flags = flags ?? new Set<string>();
    const screening = content.tables["stream_screening"];
    const order: string[] = screening.source_priority;
    for (const o of accepted) {
      // already sorted by (ts, obs_id)
      const usable = o.quality !== "artifact_likely" || o.mechanism_protected;
      if (o.type === "event") {
        this.events.push(o);
        continue;
      }
      if (!usable || !("value" in o) || o.value === undefined) {
        continue;
      }
      const prioIdx = order.indexOf(o.source);
      const prio = prioIdx >= 0 ? prioIdx : order.length;
      let point: Point;
      if (own(this.enums, o.type)) {
        const idx = this.enums[o.type].indexOf(o.value as string);
        point = [o.ts, idx, o.value as string, o.stream, prio];
      } else {
        point = [o.ts, o.value as number, null, o.stream, prio];
      }
      const arr = this.byMetric.get(o.type);
      if (arr === undefined) {
        this.byMetric.set(o.type, [point]);
      } else {
        arr.push(point);
      }
    }
  }

  ordinalIndex(metric: string, label: unknown): number | null {
    const order = own(this.enums, metric) ? this.enums[metric] : undefined;
    if (order === undefined || typeof label !== "string" || !order.includes(label)) {
      return null;
    }
    return order.indexOf(label);
  }

  inWindow(metric: string, windowMin: number): Point[] {
    const lo = this.ref - windowMin * 60;
    return (this.byMetric.get(metric) ?? []).filter((p) => lo <= p[0] && p[0] <= this.ref);
  }

  /** Stream fusion (DECISIONS.md D19): use the highest-trust single stream
   * that can answer the question on its own. Pooling readings from different
   * devices is a last resort — inter-device offsets masquerade as trends —
   * and is always flagged, never silent. */
  private select(metric: string, windowMin: number, minNeeded: number): Point[] {
    const pts = this.inWindow(metric, windowMin);
    if (pts.length === 0) {
      return pts;
    }
    const groups = new Map<string, Point[]>();
    for (const p of pts) {
      // key sorts identically to Python's (priority, stream) tuple ordering
      const key = `${p[4]}\u0000${p[3]}`;
      const arr = groups.get(key);
      if (arr === undefined) {
        groups.set(key, [p]);
      } else {
        arr.push(p);
      }
    }
    for (const key of Array.from(groups.keys()).sort()) {
      const g = groups.get(key)!;
      if (g.length >= minNeeded) {
        return g;
      }
    }
    if (groups.size > 1 && pts.length >= minNeeded) {
      this.flags.add("pooled_streams");
      return pts;
    }
    return pts;
  }

  latest(metric: string, windowMin: number): Point | null {
    const pts = this.select(metric, windowMin, 1);
    return pts.length > 0 ? pts[pts.length - 1] : null;
  }

  windowMedian(metric: string, windowMin: number, minPoints: number): number | null {
    const pts = this.select(metric, windowMin, minPoints);
    if (pts.length < minPoints) {
      return null;
    }
    return median(pts.map((p) => p[1]));
  }

  slope(metric: string, windowMin: number, minPoints: number): number | null {
    const pts = this.select(metric, windowMin, minPoints);
    return olsSlope(pts.map((p) => [p[0], p[1]] as const), minPoints);
  }

  sustainedMin(metric: string, windowMin: number, threshold: number, direction: string): number | null {
    const pts = this.select(metric, windowMin, 1);
    if (pts.length === 0) {
      return null;
    }
    const beyond = (v: number) => (direction === "above" ? v > threshold : v < threshold);
    if (!beyond(pts[pts.length - 1][1])) {
      return 0.0;
    }
    let start = pts.length - 1;
    while (start > 0 && beyond(pts[start - 1][1])) {
      start -= 1;
    }
    return q6((pts[pts.length - 1][0] - pts[start][0]) / 60.0);
  }

  crossings(metric: string, windowMin: number, threshold: number, direction: string): number | null {
    const pts = this.select(metric, windowMin, 2);
    if (pts.length < 2) {
      return null;
    }
    let count = 0;
    for (let i = 1; i < pts.length; i++) {
      const prevV = pts[i - 1][1];
      const curV = pts[i][1];
      if (direction === "above" && prevV <= threshold && threshold < curV) {
        count += 1;
      } else if (direction === "below" && prevV >= threshold && threshold > curV) {
        count += 1;
      }
    }
    return count;
  }

  eventPresent(eventId: string, windowMin: number): boolean {
    const lo = this.ref - windowMin * 60;
    return this.events.some((e) => e.event_id === eventId && lo <= e.ts && e.ts <= this.ref);
  }
}

export function classifyTrajectory(
  features: Features,
  content: ContentHandle,
  trace: Array<{ stage: string; detail: string }>,
): string {
  const rules = content.tables["trajectory_rules"];
  let worse = 0;
  let improving = 0;
  let evaluated = 0;
  const details: string[] = [];
  for (const m of rules.metrics) {
    const s = features.slope(m.metric, rules.window_min, rules.min_points);
    if (s === null) {
      continue;
    }
    evaluated += 1;
    const thr = m.slope_per_min;
    if (m.worse_direction === "up") {
      if (s >= thr) {
        worse += 1;
        details.push(`${m.metric} worsening (slope ${fmtVal(s)}/min)`);
      } else if (s <= -thr) {
        improving += 1;
      }
    } else {
      if (s <= -thr) {
        worse += 1;
        details.push(`${m.metric} worsening (slope ${fmtVal(s)}/min)`);
      } else if (s >= thr) {
        improving += 1;
      }
    }
  }
  let result: string;
  if (evaluated === 0) {
    result = "unknown";
  } else if (worse >= rules.worsening_min_signals) {
    result = "worsening";
  } else if (improving >= rules.improving_min_signals && worse === 0) {
    result = "improving";
  } else {
    result = "stable";
  }
  let detail = `trajectory ${result} (worse_signals=${worse}, improving_signals=${improving}, metrics_evaluated=${evaluated})`;
  if (details.length > 0) {
    detail += ": " + details.join("; ");
  }
  trace.push({ stage: "M4", detail });
  return result;
}
